src/ape/managers/chain.py

```diff
@@ -377,89 +383,113 @@ def _get_block(self, block_id: BlockID) -> BlockAPI:
+            yield receipt
 
-        return self._address_list_map.get(address_key, [])
+    def __iter__(self) -> Iterator[ReceiptAPI]:  # type: ignore[override]
+        yield from self.outgoing
 
-    def __iter__(self) -> Iterator[AddressType]:
-        """
-        Iterate through the accounts listed in the history map.
+    def __len__(self) -> int:
+        return len(list(self.outgoing))
 
-        Returns:
-            List[str]
-        """
+    def append(self, receipt: ReceiptAPI):
+        if receipt.txn_hash not in [x.txn_hash for x in self.sessional]:
+            self.sessional.append(receipt)
 
-        yield from self._address_list_map
+    def revert_to_block(self, block_number: int):
+        self.sessional = [x for x in self.sessional if x.block_number <= block_number]
 
-    def items(self) -> Iterator[Tuple[AddressType, List[ReceiptAPI]]]:
+
+class TransactionHistory(BaseManager):
+    """
+    A container mapping Transaction History to the transaction from the active session.
+    """
+
+    _account_history_cache: Dict[AddressType, AccountHistory] = {}
+    _hash_to_receipt_map: Dict[str, ReceiptAPI] = {}
+
+    @singledispatchmethod
+    def __getitem__(self, key):
+        raise NotImplementedError(f"Cannot use type of {type(key)} as Index")
+
+    @__getitem__.register
+    def __getitem_base_address(self, address: BaseAddress) -> AccountHistory:
+        return self.get_account_history(address)
+
+    @__getitem__.register
+    def __getitem_str(self, account_or_hash: str) -> Union[AccountHistory, ReceiptAPI]:
         """
-        Iterate through the list of address-types to list of transaction receipts.
+        Get a receipt from the history by its transaction hash.
+        If the receipt is not currently cached, will use the provider
+        to retrieve it.
+
+        Args:
+            transaction_hash (str): The hash of the transaction.
 
         Returns:
-            Iterator[Tuple[``AddressType``, :class:`~ape.api.transactions.ReceiptAPI`]]
+            :class:`~ape.api.transactions.ReceiptAPI`: The receipt.
         """
-        yield from self._address_list_map.items()
 
-    def values(self) -> Iterator[List[ReceiptAPI]]:
-        yield from self._address_list_map.values()
+        try:
+            address = self.provider.network.ecosystem.decode_address(account_or_hash)
+            return self.get_account_history(address)
+        except Exception:
+            # Use Transaction hash
+            receipt = self._hash_to_receipt_map.get(account_or_hash)
+            if not receipt:
+                # TODO: Replace with query manager once supports receipts
+                #  instead of transactions.
+                # TODO: Add timeout = 0 once in API method to not wait for txns
+                receipt = self.provider.get_receipt(account_or_hash)
+                self.append(receipt)
+
+            return receipt
+
+    @singledispatchmethod
+    def __contains__(self, key):
```

Comment:
Is there an implementation for this?

Reply:
If not, might as well leave out for now (unless it needed to have `__contains__` for some reason)

Reply:
removed!